### ISSUE TITLE
BTS Error Invalid input of type: 'dict'

### ISSUE THREAD

**Issue by bradrisse** [MKaptein/streamingbandit#81]
After setting up using docker I was able to get everything up and running. The issue is when I attempt to run the example "Bootstrapped Thompson Sampling". I run the simulation and receive the error below.

```
{
    "error": {
        "message": "Internal Server Error",
        "code": 500,
        "traceback": [
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/lib/python3.5/site-packages/tornado/web.py\", line 1697, in _execute\n    result = method(*self.path_args, **self.path_kwargs)\n",
            "  File \"/sb/app/handlers/evalhandlers.py\", line 67, in get\n    __EXP__.run_reward_code(context, action, reward)\n",
            "  File \"/sb/app/core/experiment.py\", line 131, in run_reward_code\n    exec(byte_code, {'__builtins__' : self.safe_builtins})\n",
            "  File \"\", line 21, in \n",
            "  File \"/sb/app/core/experiment.py\", line 200, in set_theta\n    return self.db.set_theta(thetas, db_key)\n",
            "  File \"/sb/app/db/database.py\", line 32, in set_theta\n    self.r_server.hmset(key, thetas)\n",
            "  File \"/usr/local/lib/python3.5/site-packages/redis/client.py\", line 2693, in hmset\n    return self.execute_command('HMSET', name, *items)\n",
            "  File \"/usr/local/lib/python3.5/site-packages/redis/client.py\", line 774, in execute_command\n    connection.send_command(*args)\n",
            "  File \"/usr/local/lib/python3.5/site-packages/redis/connection.py\", line 620, in send_command\n    self.send_packed_command(self.pack_command(*args))\n",
            "  File \"/usr/local/lib/python3.5/site-packages/redis/connection.py\", line 663, in pack_command\n    for arg in imap(self.encoder.encode, args):\n",
            "  File \"/usr/local/lib/python3.5/site-packages/redis/connection.py\", line 125, in encode\n    \"byte, string or number first.\" % typename)\n",
            "redis.exceptions.DataError: Invalid input of type: 'dict'. Convert to a byte, string or number first.\n"
        ]
    }
}
```

## Attempts to Correct:
1. Run locally
2. Use a older versions of Redis
3. Uncomment `self.params = collections.OrderedDict(self.params)` in `bts.py`
4. Remove `default_params`

**Comment by g0ulash** [comment]
Hi @bradrisse,

Can you tell me how you have set up using the Docker install? I have just run a fresh Docker image and I have no problems here.

Also running it manually with the newest Redis (5.0.5) I have no problems running the BTS example.

I'm curious what happens here, so maybe we could figure this out if you give me some more details on the steps that you took.

**Comment by Dascienz** [comment]
Hello @g0ulash,

I encountered the same error while trying to simulate the BTS experiment. The hot fix that worked for me was to specify `redis==2.10.6` in `setup.py`. I know you mentioned that you tested this with newer Redis versions, but I think there may be something strange going on there worth further investigation. 

Some of the code has been altered on my end, but nothing worked until that exact change. Here are my Dockerfile and docker-compose.yml for reference:

Dockerfile:
```
FROM python:3.5-jessie
ADD ./ /sb/
WORKDIR /sb/
RUN pip install --upgrade pip && \
    pip install .
WORKDIR /sb/app/
CMD ["python3", "./app.py"]
```

docker-compose.yml
```
version: "3"

services:

  redis:
    image: "redis:4.0.14"
    ports:
      - "6379:6379"

  mongo:
    image: "mongo:3.4"

  web:
    depends_on:
      - redis
      - mongo
    build: "."
    ports:
     - "8080:8080"
    environment:
      - "env=docker"
```

And lastly, I altered `insert_admin.py`, though this shouldn't have affected anything redis-related:
```python
    dirs = os.listdir()
    if 'app' in dirs:
        f = open('app/config.cfg', 'r')
    else:
        f = open('./config.cfg', 'r')
    settings = yaml.full_load(f) # altered to get rid of WARNING
    settings = settings['docker'] # altered to fix mongo_ip and mongo_port KeyErrors
    mongo_client = MongoClient(host=settings['mongo_ip'], port=settings['mongo_port'])
    mongo_db = mongo_client['userinfo']
    userinfo = mongo_db['userinfo']
    f.close()
```

Hope this helps!

**Comment by g0ulash** [comment]
@Dascienz thanks for checking into this!

Indeed when updating redis-py inside the Docker container I had the same conflicting errors. Also updating both redis-py and redis-server in the container did not fix the issue. For now I've created a hotfix and pushed it to the master branch. I tested your versions and it should work now as is.

For the future it's worth investigating where it goes wrong such that we can run the newest redis versions.
